Refactor Dropdowns

alexpaxton opened the pull request:
Closes #3494 

_Briefly describe your proposed changes:_
Introduce a new Dropdown component intended to replace the old component. The new component is more declarative and flexible, and embraces future-facing patterns. I have replaced some prominent dropdowns in the UI with the new component, but not all of them because there are a lot.

_What was the problem?_
There are a lot of custom built dropdowns in the UI and it makes updating the styles difficult because the markup is not consistent. The original component is not flexible enough to support all the ways it is currently being used.

_What was the solution?_

Example usage:
```
<Dropdown onChange={}>
    <DropdownItem value={} />
    <DropdownDivider />
</Dropdown>
```
I have introduced `enums` for the icon font, component colors, sizes and shapes. All the the reusable UI components follow similar customization patterns and make use of these enums.

In the future I will build a multi-select and autocomplete dropdown that will work the same way as above. This will make using any kind of dropdown interchangeable and flexible.

NOTE:
I have refactored some older components in the process of rolling out this new dropdown

## Checklist
  - [x] Rebased/mergeable
  - [x] Tests pass

jaredscheib commented on ui/src/reusable_ui/components/dropdowns/Dropdown.tsx:
```diff
@@ -0,0 +1,174 @@
+
+  private handleItemClick = (value: any) => {
+    const {onChange} = this.props
+    onChange(value)
```
what happens if whatever this `onChange` maps to, fails at any time? would you still want to collapse the menu? is it possible that the `onChange` ever maps to an async function and you may want to `await` and surround with a `try/catch`?

alexpaxton replied:
It is typed as a non-async function in props should it should prevent misuse

jaredscheib commented on ui/src/reusable_ui/components/dropdowns/Dropdown.tsx:
```diff
@@ -0,0 +1,174 @@
+import {ErrorHandling} from 'src/shared/decorators/errors'
+import './Dropdown.scss'
+
+interface Props {
```
why are there so many optional props? is this a smell of something else?

alexpaxton replied:
I would say it also reflects the modular styles used in the atomic components

jaredscheib commented on ui/src/reusable_ui/components/dropdowns/DropdownDivider.tsx:
```diff
@@ -0,0 +1,16 @@
+  <div className={classnames('dropdown--divider', {line: !text})}>{text}</div>
+)
+
+DropdownDivider.defaultProps = {
```
what do you think about type-annotating this as a `Partial<Props>`?

alexpaxton replied:
I don't know how to do it, investigated a little and couldn't find an example of doing it this way

jaredscheib commented on ui/src/reusable_ui/components/dropdowns/DropdownButton.tsx:
```diff
@@ -0,0 +1,57 @@
+
+@ErrorHandling
+class DropdownButton extends Component<Props> {
+  public static defaultProps = {
```
what do you think about type-annotating this as a `Partial<Props>`?

jaredscheib commented on ui/src/reusable_ui/components/dropdowns/DropdownButton.tsx:
```diff
@@ -0,0 +1,57 @@
+import {ErrorHandling} from 'src/shared/decorators/errors'
+import './DropdownButton.scss'
+
+interface Props {
```
why are so many of these optional?

alexpaxton replied:
Makes it easier to work with if you don't have to specify every aspect of the appearance. The defaults represent the most common appearance so you only have to specify when you want to use a non default appearance

jaredscheib commented on ui/src/reusable_ui/components/dropdowns/DropdownButton.tsx:
```diff
@@ -0,0 +1,57 @@
+    })
+  }
+
+  private get icon(): JSX.Element {
```
might this need to be that it returns `JSX.Element | null`?

jaredscheib commented on ui/src/reusable_ui/components/dropdowns/DropdownItem.tsx:
```diff
@@ -0,0 +1,61 @@
+    onClick(value)
+  }
+
+  private get checkBox(): JSX.Element {
```
does this need to be `JSX.Element | null`?

jaredscheib commented on ui/src/reusable_ui/components/dropdowns/DropdownItem.tsx:
```diff
@@ -0,0 +1,61 @@
+    )
+  }
+
+  private handleClick = () => {
```
could you type-annotate the output?

alexpaxton replied:
Done

jaredscheib commented on ui/src/reusable_ui/components/dropdowns/DropdownItem.tsx:
```diff
@@ -0,0 +1,61 @@
+    return null
+  }
+
+  private get dot(): JSX.Element {
```
does this need to be `JSX.Element | void`?

jaredscheib commented on ui/src/reusable_ui/types/index.ts:
```diff
@@ -7,6 +7,13 @@ export enum ComponentColor {
   Alert = 'alert',
 }
 
+export enum DropdownMenuColor {
```
noiiicee

jaredscheib commented on ui/src/reusable_ui/types/index.ts:
```diff
@@ -43,3 +50,71 @@ export enum Greys {
   White = '#ffffff',
 }
+
+export enum IconFont {
```
hell yeah

jaredscheib commented on ui/src/reusable_ui/types/index.ts:
```diff
@@ -43,3 +50,71 @@ export enum Greys {
+  GraphLine = 'graphline2',
+  Group = 'group',
+  Heroku = 'heroku',
+  herokusimple = '',
```
any reason this is lowercase? seems like it'd be `HerokuSimple`

alexpaxton replied:
good catch

jaredscheib commented on ui/src/reusable_ui/types/index.ts:
```diff
@@ -43,3 +50,71 @@ export enum Greys {
+  UserRemove = 'user-remove',
+  Wood = 'wood',
+  Wrench = 'wrench',
+  AlertTriangle = 'alert-triangle',
```
what do you think about making this fully alphabetically sorted? could help with later additions and general readability/dev-friendliness

alexpaxton replied:
Sure! Is there an automated way to do that?

alexpaxton commented on ui/src/hosts/containers/HostsPage.js:
```diff
@@ -5,7 +5,7 @@ import {bindActionCreators} from 'redux'
 
 import HostsTable from 'src/hosts/components/HostsTable'
-import AutoRefreshDropdown from 'shared/components/AutoRefreshDropdown'
+import AutoRefreshDropdown from 'shared/components/auto_refresh_dropdown/AutoRefreshDropdown'
```
Reusable UI is for stuff that is purely presentational, and these are built on top of the regular `Dropdown` component

alexpaxton commented on ui/src/reusable_ui/components/dropdowns/Dropdown.tsx:
```diff
@@ -0,0 +1,174 @@
+import './Dropdown.scss'
+
+interface Props {
+  children: Array<JSX.Element | JSX.Element[]>
```
It happens more than you would expect

alexpaxton replied:
Basically when you want to mix & match what appears in a dropdown. When you place a fragment in as a child it is typed as an array of elements, hence the array of arrays type here

alexpaxton commented on ui/src/reusable_ui/components/dropdowns/Dropdown.tsx:
```diff
@@ -0,0 +1,174 @@
+            maxHeight={maxMenuHeight}
+          >
+            <div className="dropdown--menu">
+              {this.flatChildren.map((child: JSX.Element) =>
```
This should theoretically never happen. I'll add a guard against it that throws an error

alexpaxton commented:
@jaredscheib 

1. Can do
2. What you are describing is how the old dropdowns worked and was one of the biggest pain points I had with using them. One of my goals with refactoring this is to make it as flexible as possible. If I wanted to create a new type of dropdown in which the items appear differently I can create a menu item component specifically for that dropdown without having to modify the Dropdown components as a whole. I considered passing in an array of objects which would include render props, but I opted for the children method as it seemed more legible and "reacty". It also seems like a pattern I want to employ in the yet to be completed reusable UI components
3. Currently it will just render an empty menu and not provide much use. My plan is to create 2 more dropdown wrappers which can all be passed the same child components (Multi-select and autocomplete dropdowns). This is still a work in progress though, so when I add the other 2 dropdown types things may change
4. Namespacing with `dropdown_` first seems nice, that way they would appear next to each other in the directory
5. the reusable UI is the temporary home for what will eventually be the UI kit, so it's exclusively for presentational components. See https://github.com/influxdata/chronograf/issues/3946 for my working list.
6. Same as point 3 -- I'm going to make 2 more dropdown types that will live in that folder. The idea is that one could change a dropdown from single to multi-select by changing the wrapper component (for example)
7. I did not know that was a thing. I'll look for an example and update

russorat commented:
can you confirm that this closes https://github.com/influxdata/chronograf/issues/3987 ?

alexpaxton commented:
@russorat I don't think that is related. That being said I have refactored the time selector dropdown in this PR so we'll see if the issue still exists. Hard to test on my own, would be great if Tim could take a look after this gets merged

alexpaxton commented on ui/src/kapacitor/components/Threshold.tsx:
```diff
@@ -60,7 +66,7 @@ class Threshold extends Component<Props> {
   }
 
-  private get operators() {
+  private get operators(): TypeItem[] {
```
I just wanted to have a return type on the getter

alexpaxton commented on ui/src/reusable_ui/components/dropdowns/Dropdown.tsx:
```diff
@@ -0,0 +1,174 @@
+          >
+            <div className="dropdown--menu">
+              {this.flatChildren.map((child: JSX.Element) =>
+                React.cloneElement(child, {
```
I updated the key to not use `uuid` so it won't re-render so much. Not sure if there's a cleaner way to handle the double-mapping process I employed during flattening

chnn replied:
We might be able to avoid cloning via component specialization. Something like:

```
// ...
<div className="dropdown-menu">
  {this.props.children.map(child => {
    if (child instanceof Dropdown.Item) {
      return (
        <Dropdown.Item
          {...child.props}
          key: `dropdown-menu--${child.props.text}`,
          selected: child.props.text === selectedItem,
          onClick: this.handleItemClick
        />
      )
    } else {
      return child
    }
  })}
</div>
```

Then hopefully this would let us avoid flattening fragments out of children, too.

alexpaxton commented on ui/src/reusable_ui/types/index.ts:
```diff
@@ -7,6 +7,13 @@ export enum ComponentColor {
   Alert = 'alert',
 }
 
+export enum DropdownMenuColors {
```
meh indeed

alexpaxton commented on ui/src/shared/components/calendar_selector/CalendarSelector.tsx:
```diff
@@ -0,0 +1,248 @@
+          <div
+            key={shortcut.id}
+            className="calendar-selector--shortcut"
+            onClick={this.handleShortcutClick(shortcut)}
```
Works in the browser

alexpaxton commented on ui/src/shared/components/dropdown_auto_refresh/AutoRefreshDropdown.tsx:
```diff
@@ -0,0 +1,95 @@
+          <Dropdown.Divider text="AutoRefresh" />
+          {autoRefreshOptions.map(option => (
+            <Dropdown.Item
+              key={option.text}
```
Tough to say -- if the dev decides to pass in children via mapping over an array they will need to assign a `key` but if they don't use map then no keys. From the perspective of the dropdown component it's hard to take any action when keys may or may not be present in its children

alexpaxton commented on ui/src/shared/components/dropdown_calendar/CalendarDropdown.tsx:
```diff
@@ -0,0 +1,113 @@
+}
+
+@ErrorHandling
+class CustomTimeRangeDropdown extends PureComponent<Props, State> {
```
Just tested the `autoRefresh` case, looks like the calender widget is unaffected as far as I can tell

alexpaxton replied:
I switched it to a regular Component

alexpaxton commented on ui/src/shared/components/dropdown_fill_query/FillQueryDropdown.tsx:
```diff
@@ -0,0 +1,118 @@
+          disabled={isDisabled}
+        >
+          {fillQueryOptions.map(option => (
+            <Dropdown.Item key={uuid.v4()} text={option.label} value={option} />
```
Done

alexpaxton commented on ui/src/shared/components/dropdown_fill_query/fillQueryOptions.ts:
```diff
@@ -0,0 +1,45 @@
+export const LINEAR = 'linear'
```
Oh yeah

alexpaxton commented on ui/src/shared/components/dropdown_query_templates/QueryTemplatesDropdown.tsx:
```diff
@@ -0,0 +1,44 @@
+        width={142}
+      >
+        {QUERY_TEMPLATES.map(option => {
+          if (option.text === DIVIDER) {
```
done!

alexpaxton commented on ui/src/shared/components/dropdown_time_range/TimeRangeDropdown.tsx:
```diff
@@ -0,0 +1,170 @@
+    super(props)
+
+    this.state = {
+      isSelectingAbsolute: false,
```
I went with `datePickerVisible`

alexpaxton commented on ui/src/shared/components/dropdown_time_range/TimeRangeDropdown.tsx:
```diff
@@ -0,0 +1,170 @@
+      return 216
+    }
+
+    return 297
```
Updated to use enum

alexpaxton commented on ui/src/shared/components/dropdown_time_shift/TimeShiftDropdown.tsx:
```diff
@@ -0,0 +1,39 @@
+      onChange={onChooseTimeShift}
+      selectedItem={selected || 'none'}
+      disabled={isDisabled}
+      width={73}
```
Updated to use constant

alexpaxton commented on ui/src/shared/components/dropdown_time_shift/TimeShiftDropdown.tsx:
```diff
@@ -0,0 +1,39 @@
+      width={73}
+    >
+      {TIME_SHIFTS.map(option => (
+        <Dropdown.Item key={uuid.v4()} text={option.label} value={option} />
```
Removed UUID

alexpaxton commented on ui/src/shared/components/dropdown_calendar/CalendarDropdown.tsx:
```diff
@@ -0,0 +1,113 @@
+    const {onApplyTimeRange} = this.props
+
+    onApplyTimeRange(timeRange)
+    this.handleCollapseCalendar()
```
Done

alexpaxton commented on ui/src/shared/components/calendar_selector/CalendarSelector.tsx:
```diff
@@ -0,0 +1,248 @@
+    return null
+  }
+
+  private getInitialDate = time => {
```
Done

alexpaxton commented on ui/src/shared/components/calendar_selector/CalendarSelector.tsx:
```diff
@@ -0,0 +1,248 @@
+
+@ErrorHandling
+class CalendarSelector extends Component<Props, State> {
+  public static defaultProps = {
```
Done

alexpaxton commented on ui/src/shared/components/calendar_selector/CalendarSelector.tsx:
```diff
@@ -0,0 +1,248 @@
+  private lowerInputRef: HTMLInputElement
+  private upperInputRef: HTMLInputElement
+  private lowerRomeRef
+  private upperRomeRef
```
Done

alexpaxton commented:
@jaredscheib I removed all the usage of UUID as keys, took care of magic numbers, and renamed some things to be more clear and declarative. Should be good!

121watts commented on ui/src/shared/components/dropdown_calendar/CalendarDropdown.tsx:
```diff
@@ -0,0 +1,113 @@
+    return `${lower} —  ${upper}`
+  }
+
+  private get buttonWidth(): number {
```
I would probably add the unit to this `buttonWidthPixels`.  It would make these numbers not be magic.

alexpaxton replied:
Nice

121watts commented on ui/src/shared/components/dropdown_time_shift/TimeShiftDropdown.tsx:
```diff
@@ -0,0 +1,44 @@
+import React, {SFC} from 'react'
+
+import Dropdown from '../../../reusable_ui/components/dropdowns/Dropdown'
```
🙅‍♂️ on these relative paths.

alexpaxton replied:
Blarghhh my text editor did a funky last night and replaced all paths with relative ones in the entire code base, I missed on apparently

alexpaxton commented:
@chnn I really like that idea. I was thinking about this last night and matching based on a text string can bring back the issue of 2 items appearing selected inside the list. Plus that allows for more flexibility when creating things like the color selector dropdowns

alexpaxton commented:
In attempting this PR I have learned that it is a drastically better process to introduce the new component (and patterns) in one PR, then replace old instances with the new in scoped PRs over time.

I have taken the feedback from this PR and opened a new PR that just introduces the new component here: https://github.com/influxdata/chronograf/pull/4041

I am going to close this PR but most likely refer back to it when forthcoming PRs are made to use the new dropdown in the UI
